index.bs

```diff
@@ -1441,55 +2040,142 @@ partial interface MLGraphBuilder {
+    : <dfn>activation</dfn>
+    ::
+            An {{MLActivation}} object.
+            Specifies the optional activation function that immediately follows the convolution operation.
+</dl>
 
-    *output size = 1 + (input size - (filter size - 1) ** *dilation - 1 + beginning padding + ending padding) / stride*
+<div>
+    **Arguments:**
+        - *input*: an {{MLOperand}}. The input 4-D tensor. The logical shape
+            is interpreted according to the value of *options*.{{MLConv2dOptions/inputLayout}}.
+        - *filter*: an {{MLOperand}}. The filter 4-D tensor. The logical shape is
+            interpreted according to the value of *options*.{{MLConv2dOptions/filterLayout}} and *options*.{{MLConv2dOptions/groups}}.
+        - *options*: an {{MLConv2dOptions}}. The optional parameters of the operation.
 
-    <div class="note">
-    A *depthwise* conv2d operation is a variant of grouped convolution, used in models like the MobileNet, where the *options.groups* = input_channels = output_channels and the shape of filter tensor is [options.groups, 1, height, width]
+    **Returns:** an {{MLOperand}}. The output 4-D tensor that contains the convolution result. The output shape is interpreted according to the *options*.{{MLConv2dOptions/inputLayout}} value. More specifically, the spatial dimensions or the sizes of the last two dimensions of the output tensor for the *nchw* input layout can be calculated as follow:
+
+    *outputSize = 1 + (inputSize - (filterSize - 1) ** *dilation - 1 + beginningPadding + endingPadding) / stride*
+</div>
+
+<div class="note">
+    A *depthwise* conv2d operation is a variant of grouped convolution, used in models like the MobileNet, where the *options.groups* = inputChannels = outputChannels and the shape of filter tensor is [options.groups, 1, height, width]
     for *"oihw"* layout, [height, width, 1, options.groups] for *"hwio"* layout, [options.groups, height, width, 1] for *"ohwi"* layout and [1, height, width, options.groups] for *"ihwo"* layout.
-    </div>
 </div>
 
+<details open algorithm>
+
+  <summary>
+    The <dfn method for=MLGraphBuilder>conv2d(|input|, |filter|, |options|)</dfn> method steps are:
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: the type of |input| and |filter| is {{MLOperand}}.
+    1. Let |inputSize| be the [=list/size=] of |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}}.
+    1. Let |filterSize| be the [=list/size=] of |filter|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}}.
+    1. If |inputSize| is not 4, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If |filterSize| is not 4, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If the type of |input| and |filter| is not the same, then [=exception/throw=] a {{TypeError}}.
```

Comment:
I believe the intention here is: _If input.[[descriptor]].type is not the same as filter.[[descriptor]].type, ..._

... but I'm not sure - "type" is ambiguous. I see this usage elsewhere: _"Assert: the type of each operand in inputs is the same."_ where "type" there is linked to MLOperandDescriptor/type. At a minimum, link "type" to that definition?

Alternately maybe define "The type of an MLOperand op is op.[[descriptor]].type" and then this line (and the other assert) could link to that. (More correct but more work.)